File: java/exercise1/DoublyLinkedList.java
package exercise1;

public class DoublyLinkedList<E> {

  private static class Node<E> {
    private E element;
    private Node<E> prev;
    private Node<E> next;

    public Node(E e, Node<E> p, Node<E> n) {
      element = e;
      prev = p;
      next = n;
    }

    public E getElement() {
      return element;
    }

    public Node<E> getPrev() {
      return prev;
    }

    public Node<E> getNext() {
      return next;
    }

    public void setPrev(Node<E> p) {
      prev = p;
    }

    public void setNext(Node<E> n) {
      next = n;
    }
  }

  private Node<E> header;
  private Node<E> trailer;
  private int size = 0;

  public DoublyLinkedList() {
    header = new Node<>(null, null, null);
    trailer = new Node<>(null, header, null);
    header.setNext(trailer);
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public E first() {
    if (isEmpty())
      return null;
    return header.getNext().getElement();
  }

  public E last() {
    if (isEmpty())
      return null;
    return trailer.getPrev().getElement();
  }

  public void addFirst(E e) {
    addBetween(e, header, header.getNext());
  }

  public void addLast(E e) {
    addBetween(e, trailer.getPrev(), trailer);
  }

  public E removeFirst() {
    if (isEmpty())
      return null;
    return remove(header.getNext());
  }

  public E removeLast() {
    if (isEmpty())
      return null;
    return remove(trailer.getPrev());
  }

  private void addBetween(E e, Node<E> predecessor, Node<E> successor) {
    Node<E> newest = new Node<>(e, predecessor, successor);
    predecessor.setNext(newest);
    successor.setPrev(newest);
    size++;
  }

  private E remove(Node<E> node) {
    Node<E> predecessor = node.getPrev();
    Node<E> successor = node.getNext();
    predecessor.setNext(successor);
    successor.setPrev(predecessor);
    size--;
    return node.getElement();
  }

  public String toString() {
    StringBuilder sb = new StringBuilder("(");
    Node<E> walk = header.getNext();
    while (walk != trailer) {
      sb.append(walk.getElement());
      walk = walk.getNext();
      if (walk != trailer)
        sb.append(", ");
    }
    sb.append(")");
    return sb.toString();
  }

  public void swapTwoNodes(Node<E> node1, Node<E> node2) {
    if (node1 == node2)
      return;

    Node<E> prev1 = node1.getPrev();
    Node<E> next1 = node1.getNext();
    Node<E> prev2 = node2.getPrev();
    Node<E> next2 = node2.getNext();

    if (node1 == next2) {
      // node1 is right after node2
      node2.setNext(node1);
      node2.setPrev(prev1);
      node1.setPrev(node2);
      node1.setNext(next1);
      if (prev1 != null)
        prev1.setNext(node2);
      if (next1 != null)
        next1.setPrev(node1);
    } else if (node2 == next1) {
      // node2 is right after node1
      node1.setNext(node2);
      node1.setPrev(prev2);
      node2.setPrev(node1);
      node2.setNext(next2);
      if (prev2 != null)
        prev2.setNext(node1);
      if (next2 != null)
        next2.setPrev(node2);
    } else {
      // nodes are not adjacent
      node1.setNext(next2);
      node1.setPrev(prev2);
      node2.setNext(next1);
      node2.setPrev(prev1);
      if (prev1 != null)
        prev1.setNext(node2);
      if (next1 != null)
        next1.setPrev(node2);
      if (prev2 != null)
        prev2.setNext(node1);
      if (next2 != null)
        next2.setPrev(node1);
    }

    // Update head and tail if necessary
    if (header.getNext() == node1)
      header.setNext(node2);
    else if (header.getNext() == node2)
      header.setNext(node1);
    if (trailer.getPrev() == node1)
      trailer.setPrev(node2);
    else if (trailer.getPrev() == node2)
      trailer.setPrev(node1);
  }

  public static void main(String[] args) {
    DoublyLinkedList<String> list = new DoublyLinkedList<>();
    list.addFirst("MSP");
    list.addLast("ATL");
    list.addLast("BOS");
    list.addFirst("LAX");

    System.out.println("Original list: " + list);

    Node<String> node1 = list.header.getNext().getNext(); // MSP
    Node<String> node2 = list.header.getNext().getNext().getNext().getNext(); // BOS
    list.swapTwoNodes(node1, node2);
    System.out.println("After swapping MSP and BOS: " + list);
  }
}

File: src/exercise1/doubly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.previous = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

        # TO check weather list is empty or not?

    def is_empty(self):
        return self.head is None

    def add_node(self, data):
        #  Creating New Node
        new_node = Node(data)

        # Checking where the head and tail pointer is pointing.
        # if first node is added to the list then below condition should be
        # true.
        if self.head is None:
            self.head = new_node
            new_node.previous = new_node.next = None

        else:
            # Addding new node at the last
            self.tail.next = new_node
            new_node.previous = self.tail
            new_node.next = None
        self.tail = new_node

    def display_list(self):
        # Creating temp pointer to traverse
        temp = self.head
        if temp:
            while temp:
                print(temp.data, end=" -> ")
                temp = temp.next
            print("NULL")
        else:
            print("List does not have any nodes")

    def swap_two_nodes(self, node1, node2):
        """
        Swaps two nodes in a doubly linked list.
        """
        # No need to swap if the nodes are the same
        if node1 is node2:
            return

        # Adjust the previous pointers of the nodes following node1 and node2.
        if node1.next:
            node1.next.previous = node2
        if node2.next:
            node2.next.previous = node1

        # Swap the next pointers of node1 and node2.
        node1.next, node2.next = node2.next, node1.next

        # Adjust the next pointers of the nodes preceding node1 and node2.
        if node1.previous:
            node1.previous.next = node2
        if node2.previous:
            node2.previous.next = node1

        # Swap the previous pointers of node1 and node2.
        node1.previous, node2.previous = node2.previous, node1.previous

        # Update the head of the list if necessary.
        if self.head == node1:
            self.head = node2
        elif self.head == node2:
            self.head = node1

        # Update the tail of the list if necessary.
        if self.tail == node1:
            self.tail = node2
        elif self.tail == node2:
            self.tail = node1


def clone_linked_list(l1):
    dl = DoublyLinkedList()
    temp = l1.head
    if temp is not None:
        while temp is not None:
            dl.add_node(temp.data)
            temp = temp.next
    return dl
